devtools/maven/src/main/java/io/quarkus/maven/MavenProjectBuildFile.java

```diff
@@ -0,0 +1,195 @@
+            }
+        } else if (model().getDependencies()
+                .stream()
+                .noneMatch(thisDep -> d.getManagementKey().equals(thisDep.getManagementKey()))) {
+            model().addDependency(d);
+            // it could still be a transitive dependency or inherited from the parent
+            if (!getDependencies().contains(coords)) {
+                getDependencies().add(coords);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    protected void removeDependency(AppArtifactKey key) throws IOException {
+        if (model() != null) {
+            final Iterator<AppArtifactCoords> i = getDependencies().iterator();
+            while (i.hasNext()) {
+                final AppArtifactCoords a = i.next();
+                if (a.getKey().equals(key)) {
+                    i.remove();
+                    break;
+                }
+                model().getDependencies().removeIf(d -> Objects.equals(toKey(d), key));
+            }
+        }
+    }
+
+    @Override
+    protected List<AppArtifactCoords> getDependencies() {
+        if (dependencies == null) {
+            final List<org.eclipse.aether.graph.Dependency> projectDeps = projectDepsSupplier.get();
+            dependencies = new ArrayList<>(projectDeps.size());
+            for (org.eclipse.aether.graph.Dependency dep : projectDeps) {
+                org.eclipse.aether.artifact.Artifact a = dep.getArtifact();
+                dependencies.add(new AppArtifactCoords(a.getGroupId(), a.getArtifactId(), a.getClassifier(),
+                        a.getExtension(), a.getVersion()));
+            }
+        }
+        return dependencies;
+    }
+
+    protected List<AppArtifactCoords> getManagedDependencies() {
+        if (managedDependencies == null) {
+            final List<org.eclipse.aether.graph.Dependency> managedDeps = projectManagedDepsSupplier.get();
+            managedDependencies = new ArrayList<>(managedDeps.size());
+            for (org.eclipse.aether.graph.Dependency dep : managedDeps) {
+                org.eclipse.aether.artifact.Artifact a = dep.getArtifact();
+                managedDependencies.add(new AppArtifactCoords(a.getGroupId(), a.getArtifactId(), a.getClassifier(),
+                        a.getExtension(), a.getVersion()));
+            }
+        }
+        return dependencies;
+    }
+
+    @Override
+    protected void writeToDisk() throws IOException {
+        if (model == null) {
+            return;
+        }
+        try (ByteArrayOutputStream pomOutputStream = new ByteArrayOutputStream()) {
+            MojoUtils.write(model(), pomOutputStream);
+            writeToProjectFile(BuildTool.MAVEN.getDependenciesFile(), pomOutputStream.toByteArray());
+        }
+    }
+
+    @Override
+    protected String getProperty(String propertyName) {
+        return projectPropsSupplier.getProperty(propertyName);
+    }
+
+    @Override
+    protected void refreshData() {
```

Comment:
Are the suppliers reading the file each time they are called?

Reply:
They are supposed to be invoked once to initialize the fields on demand. But I see I didn't do this for the properties.

Reply:
Just mind that that they have to be up to date after using `install/uninstall`

Reply:
The state of the BuildFile will be up to date. Otherwise the maven will have to be re-launched.

Reply:
Ah, `private Properties projectPropsSupplier;` is not actually a supplier. I should just rename the field.